datawig/hpo.py

```diff
@@ -0,0 +1,400 @@
+                       num_epochs=global_parms['num_epochs'],
+                       patience=global_parms['patience'],
+                       test_split=.1,
+                       weight_decay=global_parms['weight_decay'],
+                       batch_size=global_parms['batch_size'],
+                       final_fc_hidden_units=global_parms['final_fc_hidden_units'],
+                       calibrate=True)
+
+        # add suitable metrics to hp series
+        imputed = hp_imputer.predict(test_df)
+        true = imputed[simple_imputer.output_column]
+        predicted = imputed[simple_imputer.output_column + '_imputed']
+
+        imputed_train = hp_imputer.predict(train_df.sample(min(train_df.shape[0], int(1e4))))
+        true_train = imputed_train[simple_imputer.output_column]
+        predicted_train = imputed_train[simple_imputer.output_column + '_imputed']
+
+        if is_numeric_dtype(train_df[simple_imputer.output_column]):
+            hp['mse'] = mean_squared_error(true, predicted)
+            hp['mse_train'] = mean_squared_error(true_train, predicted_train)
+            confidence = float('nan')
+        else:
+            confidence = imputed[simple_imputer.output_column + '_imputed_proba']
+            confidence_train = imputed_train[simple_imputer.output_column + '_imputed_proba']
+            hp['f1_micro'] = f1_score(true, predicted, average='micro')
+            hp['f1_macro'] = f1_score(true, predicted, average='macro')
+            hp['f1_weighted'] = f1_score(true, predicted, average='weighted')
+            hp['f1_weighted_train'] = f1_score(true_train, predicted_train, average='weighted')
+            hp['precision_weighted'] = f1_score(true, predicted, average='weighted')
+            hp['precision_weighted_train'] = f1_score(true_train, predicted_train, average='weighted')
+            hp['recall_weighted'] = recall_score(true, predicted, average='weighted')
+            hp['recall_weighted_train'] = recall_score(true_train, predicted_train, average='weighted')
+            hp['coverage_at_90'] = (confidence > .9).mean()
+            hp['coverage_at_90_train'] = (confidence_train > .9).mean()
+            hp['true_precision_at_90'] = (predicted[confidence > .9] == true[confidence > .9]).mean()
+            hp['ece_pre_calibration'] = hp_imputer.calibration_info['ece_post']
+            hp['ece_post_calibration'] = hp_imputer.calibration_info['ece_post']
+
+        for uds in user_defined_scores:
+            hp[uds[1]] = uds[0](true=true, predicted=predicted, confidence=confidence)
+
+        hp_imputer.save()
+
+        return hp
+
+    def tune(self,
+             train_df: pd.DataFrame,
+             test_df: pd.DataFrame = None,
+             hps: dict = None,
+             strategy: str = 'random',
+             num_evals: int = 10,
+             max_running_hours: float = 96,
+             user_defined_scores: list = None,
+             hpo_run_name: str = '',
```

Comment:
default here could be `None` to indicate it is optional instead of an empty string.